x-pack/platform/plugins/shared/onechat/server/services/tools/esql/utils/execute_esql_query.ts

```
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EsqlTool } from '@kbn/onechat-server';
+import { z } from '@kbn/zod';
+import { EsqlToolCreateResponse } from '../../../../../common/tools';
+
+export const esqlSchema = z.object({
+  params: z
+    .record(z.string())
+    .optional()
+    .describe('Params needed to execute the query - must match defined tool parameter keys'),
+});
+
+export const esqlToolCreator = (tool: EsqlToolCreateResponse): EsqlTool => {
+  const executableTool: EsqlTool = {
+    id: tool.id,
+    name: tool.name,
+    description: tool.description,
+    query: tool.query,
+    params: tool.params,
+    schema: esqlSchema,
+    handler: async ({ params }, { esClient }) => {
+      const client = esClient.asCurrentUser;
+      const filledQuery = tool.query.replace(/\?(\w+)/g, (_, key) => {
+        if (!(key in tool.params)) {
+          throw new Error(
+            `Error with query: Parameter ${key} not found in tool params. Available: ${Object.keys(
+              tool.params
+            ).join(', ')}`
+          );
+        }
+        const value = params[key];
+        if (value === undefined || value === null) {
+          throw new Error(`Error with query: Parameter ${key} is required but was not provided`);
+        }
+
+        return typeof value === 'string' ? `"${value.replace(/"/g, '\\"')}"` : value;
+      });
+
+      const response = await client.esql.query({
+        query: filledQuery,
+      });
+
+      return response;
```

Are we fine with the default ES|QL tabular response format as output format for our tools (genuinely asking, I asked myself the same question for the built-in `execute_esql` tool)?   

fwiw, I have a utility to convert this to JSON format, if we want to use it: https://github.com/elastic/kibana/blob/04b294e67d51660b3e01fa7f61bb0492b3e6ebf9/x-pack/platform/packages/shared/onechat/onechat-genai-utils/tools/utils/esql.ts#L11-L27

---

Hmm I also thought about this so I added a [Tool Response section](https://docs.google.com/document/d/10ZlcUZhrnxpWc6NDgv1YIXIUxc6AtNVhTpto_yQl9PU/edit?tab=t.6mo83zpjv1um#heading=h.51mitiwdn5jw) to the RFC. I imagine it should be standardized so I proposed something like this?

```
{
	success: boolean,
	results?: any, // results from ESQL query
	error?: {
		code: string
		message: string
	},
	meta: {
		execution_time: number,
		provider_id: string,
		tool_id: number,
	}
		
}
```

---

So error format will be handled internally by onechat. From the ES|QL tool's standpoint, you should just throw an error.
Same for meta, I don't think we need it right now, but if/when we do, it will be standardized in the onechat framework (e.g it doesn't make sense for your tool to return `toolId` and `providerId` - the framework knows which tool it is calling and can add it to the tool's response if we want to - same goes for execution time).

So I would just return the results